diff --git a/pfetch b/pfetch
index 3f62187..fbfd778 100755
--- a/pfetch
+++ b/pfetch
@@ -2236,3 +2236,15 @@
             ;;
 
+            ([Rr]unix*)
+                read_ascii 2 <<-EOF
+					${c7}  \\  ${c2}\\ //
+					${c7} ==\\__${c2}\\/ ${c7}//
+					${c2}   //   \\${c7}//
+					${c2}==//     ${c7}//==
+					${c2} //${c7}\\${c2}___${c7}//
+					${c2}// ${c7}/\\  ${c2}\\==
+					  ${c7}// \\  ${c2}\\
+				EOF
+            ;;
+
             ([Nn]ix[Oo][Ss]*)
*** End Patch
